### ISSUE TITLE
symbolic fails lookup if there are no line records

### ISSUE THREAD

**Issue by willkg** [getsentry/symbolic#284]
With symbolic 7.5.0 and this sym file:

```
MODULE windows x86_64 09F9D7ECF31F60E34C4C44205044422E1 xul.pdb
INFO CODE_ID 5E8F310E69D0000 xul.dll
FILE 0 hg:hg.mozilla.org/mozilla-central:vs2017_15.8.4/VC/include/functional:93bdbca5399c12b3eec5f03bbc323e00f7ef3a51
FUNC c6db60 f 0 std::_Func_impl_no_alloc<`lambda at /builds/worker/checkouts/gecko/netwerk/protocol/http/HttpChannelChild.cpp:411:7',void>::_Move(void*)
c6db60 f 1219 0
FUNC c6db70 23e 0 std::_Func_impl_no_alloc<`lambda at /builds/worker/checkouts/gecko/netwerk/protocol/http/HttpChannelChild.cpp:411:7',void>::_Do_call()
FUNC c6ddb0 8 0 std::_Func_impl_no_alloc<`lambda at /builds/worker/checkouts/gecko/netwerk/protocol/http/HttpChannelChild.cpp:411:7',void>::_Target_type() const
c6ddb0 8 1227 0
FUNC c6ddc0 209 0 std::_Func_impl_no_alloc<`lambda at /builds/worker/checkouts/gecko/netwerk/protocol/http/HttpChannelChild.cpp:411:7',void>::_Delete_this(bool)
c6ddc0 b 1239 0
c6ddcb 102 1240 0
c6decd 8 1241 0
c6ded5 10 1243 0
```

this Python code fails:

```
import symbolic

debug_id = symbolic.normalize_debug_id("09F9D7ECF31F60E34C4C44205044422E1")
archive = symbolic.Archive.open("xul.sym")
obj = archive.get_object(debug_id=debug_id)
symcache = obj.make_symcache()
lineinfo = symcache.lookup("0xc6dd98")

print(lineinfo)                         
```

lineinfo ends up being an empty list because symbolic can't resolve the address. I think it's because there are no line records for that function.

I think one of the following is true:

1. our sym file is malformed--maybe there's a rule that you must have line records for a FUNC?
2. symbolic should handle this case and return what information it can for that symbol

Any ideas? Thank you!

**Comment by willkg** [comment]
For some context, I'm rewriting the symbolication API for Tecken (https://tecken.readthedocs.io/) and I want to switch from the parser we have to symbolic. Then our efforts can go towards improving symbolic instead of continuing our parser and it's another step towards switching to the Rust-based tooling for breakpad things.

**Comment by willkg** [comment]
Possibly related to this, symbolic will also fail if the address is covered by a PUBLIC record which also has no line records.

Here's the relevant parts of the sym file:

```
MODULE Linux x86_64 5A5F32914295A4FA22B591670A49FBEF0 libgallium_dri.so
INFO CODE_ID 91325F5A9542FAA422B591670A49FBEF719A143E
PUBLIC 135380 0 __driDriverGetExtensions_swrast
PUBLIC 1353a0 0 __driDriverGetExtensions_kms_swrast
PUBLIC 1353c0 0 __driDriverGetExtensions_iris
PUBLIC 1353e0 0 __driDriverGetExtensions_nouveau
PUBLIC 135400 0 __driDriverGetExtensions_r300
PUBLIC 135420 0 __driDriverGetExtensions_r600
PUBLIC 135440 0 __driDriverGetExtensions_radeonsi
PUBLIC 135460 0 __driDriverGetExtensions_vmwgfx
PUBLIC 135480 0 __driDriverGetExtensions_virtio_gpu
PUBLIC 1354a0 0 __driDriverGetExtensions_zink
PUBLIC 6dd790 0 radeon_drm_winsys_create
PUBLIC 872f20 0 amdgpu_winsys_create
PUBLIC 87cb40 0 ac_init_shared_llvm_once
PUBLIC 88e650 0 nouveau_drm_screen_create
STACK CFI INIT 103020 10 .cfa: $rsp 16 + .ra: .cfa -8 + ^ 
STACK CFI 103026 .cfa: $rsp 24 +
STACK CFI INIT 103030 8 .cfa: $rsp 8 + .ra: .cfa -8 + ^ 
STACK CFI INIT 103040 13 $r12: .cfa -24 + ^ $r13: .cfa -16 + ^ $rbp: .cfa -32 + ^ $rbx: .cfa -40 + ^ .cfa: $rsp 96 + .ra: .cfa -8 + ^ 
...
```

Given `0x1489adf`, symbolic will fail the lookup, but breakpad stackwalker returns `nouveau_drm_screen_create`.

**Comment by gabrielesvelto** [comment]
FYI breakpad's stackwalker will pick the nearest public symbol if the address falls in a range not covered by any function

**Comment by jan-auer** [comment]
Sorry for the inactivity on this part. 

I think the first example (libxul) is indeed technically invalid. We currently discard all functions without line records based on the assumption that without lines they could not be hit. Do you have the original PDB so we can check if this is a problem of dump_syms? 

We made this optimization when dealing with DWARF since we found eliminated inline functions and even eliminated functions without line records. There are probably other ways to detect this, too, which I would have to look into.

---

As for the second example (libgallium_dri), this is actually a bug in our symcache:

https://github.com/getsentry/symbolic/blob/86279cb3206d507fee4a9b68e362225cd4fe1df9/symbolic-symcache/src/format.rs#L218

This assumes that functions never get larger than 65k, but in the case of `PUBLIC` records (or, rather symbol table entries without a size), symbolic still needs to respond with the previous record. 

That said, I believe that here again the sym file may be missing entries. It's hard to imagine that `nouveau_drm_screen_create` is 12.5MB in code size.

**Comment by gabrielesvelto** [comment]
> That said, I believe that here again the sym file may be missing entries. It's hard to imagine that `nouveau_drm_screen_create` is 12.5MB in code size.

Yes, that's the typical case where we only have the executable but not the debuginfo, so the symbols we dump are limited. We encounter this type of scenario on Android and on Linux distros that don't have separate debug packages such as Arch.

**Comment by willkg** [comment]
> Do you have the original PDB so we can check if this is a problem of dump_syms?

The file is here: https://symbols.mozilla.org/xul.pdb/09F9D7ECF31F60E34C4C44205044422E1/xul.sym

@gabrielesvelto : That sym file was uploaded 10 months ago. I don't know what build it was. Any idea if the sym file has a problem we've fixed in `dump_syms`. If you think it might, I should update my example and/or see if it's still valid.

**Comment by jan-auer** [comment]
There's also this open PR in the Rust rewrite: https://github.com/mozilla/dump_syms/pull/197

Looking at that from the context of this, it may be safer to create `PUBLIC` records instead (although that would be strange considering the semantics of that record). I tried to check if there's any expectation for having line records with `FUNC`, and it seems there is none.

**Comment by gabrielesvelto** [comment]
> Looking at that from the context of this, it may be safer to create `PUBLIC` records instead (although that would be strange considering the semantics of that record). I tried to check if there's any expectation for having line records with `FUNC`, and it seems there is none.

We've encountered `FUNC` without lines in Rust mostly. rustc doesn't always produce good debuginfo and we've seen it manifests itself this way.

**Comment by jan-auer** [comment]
I see, thank you. In that case, the TODO for this issue is:

- [ ] Support public symbols > 65k in the SymCache reader.
- [ ] Do not remove functions without line records in the SymCache writer.

**Comment by jan-auer** [comment]
This has been fixed in #320 and #323 and will be inlcuded in the next bugfix release. Thanks for reporting!

**Comment by willkg** [comment]
Thank you!
